<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Raphaël Scherrer" />

<meta name="date" content="2020-04-20" />

<title>egssimtools: tools for analyzing simulation data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">egssimtools: tools for analyzing simulation data</h1>
<h4 class="author">Raphaël Scherrer</h4>
<h4 class="date">2020-04-20</h4>



<p>This package provides tools to facilitate the analysis and processing of the data saved by the simulation program ExplicitGenomeSpeciation (<a href="https://github.com/rscherrer/ExplicitGenomeSpeciation">https://github.com/rscherrer/ExplicitGenomeSpeciation</a>). It comes as part of the program repository itself. Here we show how to use it. The repository has a folder where an example simulation was already run, so we will use these data to illustrate the workings of the package. Note that you can choose what variables to save when running the simulations, and so all data files may not be present if you specified otherwise. Here, we assume that all the data are available.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(egssimtools)</span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co"># First set the simulation folder</span></span>
<span id="cb1-4"><a href="#cb1-4"></a>simulation &lt;-<span class="st"> &quot;../build/release&quot;</span></span></code></pre></div>
<p>You can easily read a list of parameter values for a given simulation by using</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>pars &lt;-<span class="st"> </span><span class="kw">read_parameters</span>(simulation)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">head</span>(pars)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; $rdynamics</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt; [1] &quot;1&quot;</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt; </span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">#&gt; $capacity</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; [1] &quot;1&quot;</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt; </span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt; $replenish</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#&gt; [1] &quot;2375&quot;</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt; </span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt; $inflow</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt; [1] &quot;400&quot;</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt; </span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt; $outflow</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt; [1] &quot;100&quot;</span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co">#&gt; </span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="co">#&gt; $hsymmetry</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co">#&gt; [1] &quot;0.75&quot;</span></span></code></pre></div>
<p>which assumes that the parameter file is called “paramlog.txt”. You can change this with the <code>filename</code> argument of the function. The parameters are returned as character strings because some of them are not numbers, and as a list because some parameters have more than one value. This returns all the parameters of the simulation, but you can also choose specific parameters to return:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">read_parameters</span>(simulation, <span class="dt">parnames =</span> <span class="kw">c</span>(<span class="st">&quot;hsymmetry&quot;</span>, <span class="st">&quot;ecosel&quot;</span>))</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt; $hsymmetry</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; [1] &quot;0.75&quot;</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; </span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; $ecosel</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; [1] &quot;1.4&quot;</span></span></code></pre></div>
<p>Here we provide functions to read variables through time. Different kinds of data are saved during the simulations, with different dimensions. All of the functions we provide here wrap around the <code>read_binary</code> function, which can read binary files, and all of them assume that “time.dat” is available (which is read using the <code>read_time</code> function). Functions for individual-level variables assume that “population_size.dat” was saved (population size is read using the <code>read_population_size</code> function). See the README for more details on how to choose what variables are possible to save.</p>
<p>To read population-wide variables use:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">read_population</span>(simulation, <span class="dt">variable =</span> <span class="st">&quot;EI&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt;  [1] 0.6219105 0.6670805 0.6725296 0.6806327 0.7137775 0.7247432 0.7273780</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt;  [8] 0.7598032 0.7856653 0.7963782 0.8121315 0.8357514 0.8196376 0.8275657</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt; [15] 0.8549591 0.8682196 0.8821501 0.8810706 0.8856761 0.8989007</span></span></code></pre></div>
<p>for ecological divergence through time, for example.</p>
<p>Alternatively, you may want to read individual-level data. For this use:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>traits &lt;-<span class="st"> </span><span class="kw">read_individuals</span>(simulation, <span class="st">&quot;individual_trait&quot;</span>, <span class="dt">nvalues =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>which returns a nested list of individual values (three values per individual in the case of trait values) clustered by time points.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># Trait values of the first individual in the first time point</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>traits[[<span class="dv">1</span>]][[<span class="dv">1</span>]]</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt; [1] -0.9936792 -0.2327907  0.8127390</span></span></code></pre></div>
<p>The argument <code>nvalues</code> is three by default, but can be changed if reading other individual-level variables such as “ecotype” or “habitat”. However, we have simplified functions for those, that return the data in the same way:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>ecotypes &lt;-<span class="st"> </span><span class="kw">read_ecotypes</span>(simulation)</span>
<span id="cb7-2"><a href="#cb7-2"></a>habitats &lt;-<span class="st"> </span><span class="kw">read_habitats</span>(simulation)</span>
<span id="cb7-3"><a href="#cb7-3"></a>ecotypes[[<span class="dv">1</span>]][[<span class="dv">1</span>]]</span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; [1] 1</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>habitats[[<span class="dv">1</span>]][[<span class="dv">1</span>]]</span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; [1] 1</span></span></code></pre></div>
<p>We can also read locus-specific variables across the genome, for example:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>loci &lt;-<span class="st"> </span><span class="kw">read_loci</span>(simulation, <span class="st">&quot;genome_Fst&quot;</span>)</span></code></pre></div>
<p>which returns a list of vectors of locus-specific values for each time point.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">head</span>(loci[[<span class="dv">1</span>]]) <span class="co"># at the first time point</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">#&gt; [1] 0.0003611217 0.0007448260 0.0052053908 0.0013079429 0.0019458267</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co">#&gt; [6] 0.0015416331</span></span></code></pre></div>
<p>Alternatively we can look at edge-specific variables in a network, with</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>edges &lt;-<span class="st"> </span><span class="kw">read_edges</span>(simulation, <span class="st">&quot;network_corgen&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="kw">str</span>(edges[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt; List of 3</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt;  $ 0  :List of 3</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt;   ..$ 0: num [1:100] 0.0863 0.0236 0.0513 -0.0404 -0.0124 ...</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt;   ..$ 1: num [1:100] -0.01064 -0.0425 -0.00965 -0.04542 -0.00281 ...</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt;   ..$ 2: num [1:100] 0.0592 0.0157 0.0373 -0.0793 -0.0159 ...</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt;  $ 100:List of 3</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt;   ..$ 0: num [1:100] 0.027625 -0.005084 -0.036151 0.040351 0.000195 ...</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt;   ..$ 1: num [1:100] -0.01581 -0.02762 -0.02222 -0.00305 0.04552 ...</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co">#&gt;   ..$ 2: num [1:100] 0.022 0.029 -0.0595 0.0246 -0.0129 ...</span></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="co">#&gt;  $ 200:List of 3</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="co">#&gt;   ..$ 0: num [1:100] 0.0758 -0.0152 0.0188 0.0587 0.0308 ...</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="co">#&gt;   ..$ 1: num [1:100] -0.01095 -0.03122 -0.00636 -0.01285 0.03684 ...</span></span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt;   ..$ 2: num [1:100] -0.00792 -0.01073 0.05337 -0.00989 0.05332 ...</span></span></code></pre></div>
<p>which returns a vector of edge-specific values for each trait, for each time point.</p>
<p>Each simulation has a genetic architecture, which is implemeted as its own class in the C++ code. If you chose to save the genetic architecture into a file (see README on how to do that), you can retrieve the information of that file with</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>arch &lt;-<span class="st"> </span><span class="kw">read_architecture</span>(simulation)</span></code></pre></div>
<p>which assumes that the architecture file name is “architecture.txt”, but you can change that using the <code>filename</code> argument. The genetic architecture is stored as a list that matches the organization of the file (explained in the README):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">str</span>(arch)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co">#&gt; List of 6</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">#&gt;  $ chromosomes: num [1:3] 0.333 0.667 1</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#&gt;  $ traits     : num [1:300] 0 1 0 0 0 2 0 2 1 1 ...</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">#&gt;  $ locations  : num [1:300] 0.000644 0.001195 0.008408 0.008765 0.009228 ...</span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">#&gt;  $ effects    : num [1:300] 0.1279 -0.0838 0.0968 -0.0596 0.0642 ...</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">#&gt;  $ dominances : num [1:300] 0.02293 0.00548 0.23688 0.02891 0.17171 ...</span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">#&gt;  $ networks   :List of 3</span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">#&gt;   ..$ ecotrait:List of 3</span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">#&gt;   .. ..$ edges0 : num [1:100] 221 221 123 123 221 221 221 221 221 183 ...</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">#&gt;   .. ..$ edges1 : num [1:100] 123 183 183 150 150 288 223 3 34 143 ...</span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">#&gt;   .. ..$ weights: num [1:100] 0.0436 0.1317 -0.1225 -0.0981 0.0392 ...</span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="co">#&gt;   ..$ matepref:List of 3</span></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co">#&gt;   .. ..$ edges0 : num [1:100] 1 88 88 206 88 1 292 88 292 292 ...</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co">#&gt;   .. ..$ edges1 : num [1:100] 88 206 292 118 118 118 118 232 154 229 ...</span></span>
<span id="cb12-16"><a href="#cb12-16"></a><span class="co">#&gt;   .. ..$ weights: num [1:100] -0.0547 -0.0787 0.0886 0.0991 0.0965 ...</span></span>
<span id="cb12-17"><a href="#cb12-17"></a><span class="co">#&gt;   ..$ neutral :List of 3</span></span>
<span id="cb12-18"><a href="#cb12-18"></a><span class="co">#&gt;   .. ..$ edges0 : num [1:100] 226 226 226 55 226 226 80 259 259 55 ...</span></span>
<span id="cb12-19"><a href="#cb12-19"></a><span class="co">#&gt;   .. ..$ edges1 : num [1:100] 236 66 55 259 80 126 126 126 76 76 ...</span></span>
<span id="cb12-20"><a href="#cb12-20"></a><span class="co">#&gt;   .. ..$ weights: num [1:100] 0.1611 -0.1062 -0.0356 -0.0752 0.1822 ...</span></span></code></pre></div>
<p>where the first five slots are lists of values across loci, and the last slot contains edge-specific values for the gene networks, sorted by network.</p>
<p>Some simulations may have crashed or not even started. To check if data are unavailable for a simulation use:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">is_missing</span>(simulation)</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>and to check if the simulation went extinct before completion use:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">is_extinct</span>(simulation)</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>You can also use wrap-arounds of those functions to find out whether and which simulations went missing or extinct in a folder containing many simulations.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">find_extinct</span>(<span class="dt">root =</span> <span class="st">&quot;../build&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;release&quot;</span>, <span class="dt">pb =</span> <span class="ot">FALSE</span>)</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co">#&gt; NULL</span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="kw">find_missing</span>(<span class="dt">root =</span> <span class="st">&quot;../build&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;release&quot;</span>, <span class="dt">pb =</span> <span class="ot">FALSE</span>)</span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>where <code>root</code> is where all the simulation folders are, <code>pattern</code> is an optional pattern-matching string to find the simulation folders (defaults to “sim_” at the beginning of the simulation folder name) and <code>pb</code> controls whether a progress bar is displayed while the function loops through the simulations. Both functions return a vector of simulation folder names where the simulation went missing or extinct, otherwise they return <code>NULL</code>.</p>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
